Modules/_datetimemodule.c

```diff
@@ -2832,19 +2832,26 @@ date_today(PyObject *cls, PyObject *dummy)
     return result;
 }
 
-/* Return new date from given timestamp (Python timestamp -- a double). */
+/*[clinic input]
+@classmethod
+datetime.date.fromtimestamp
+
+    timestamp: 'O'
+    /
+
+Create a date from a POSIX timestamp.
+
+The timestamp must be a double, e.g. created via time.time() and is
+interpreted as local time.
+[clinic start generated code]*/
+
 static PyObject *
-date_fromtimestamp(PyObject *cls, PyObject *args)
+datetime_date_fromtimestamp(PyTypeObject *type, PyObject *timestamp)
+/*[clinic end generated code: output=fd045fda58168869 input=8b4945faabab9dba]*/
 {
-    PyObject *timestamp;
-    PyObject *result = NULL;
-
-    if (PyArg_ParseTuple(args, "O:fromtimestamp", &timestamp))
-        result = date_local_from_object(cls, timestamp);
-    return result;
+    return date_fromtimestamp((PyObject *) type, timestamp);
```

Comment:
Wouldn't be better to just move the Argument Clinic declaration before `date_fromtimestamp`, and rename the later to `datetime_date_fromtimestamp`?

Reply:
Yes. I sort of just copied that from the original comment ("Python timestamp -- a double"). That's probably not reasonable anyway since this is the docstring for the Python API and thus should not use a C type. Is "The timestamp is a number, e.g. ..." ok?

Reply:
I meant that since `date_fromtimestamp` is used in just a single place, why not inline it? But instead of moving the body of `date_fromtimestamp` inside `datetime_date_fromtimestamp`, it is better to move the header of `datetime_date_fromtimestamp` (generated by Argument Clinic, together with the prepending Argument Clinic declaration) in place of the header of `date_fromtimestamp`. Both `datetime_date_fromtimestamp` and `date_fromtimestamp` are C functions, with virtually identical signatures.

Sorry if I'm not clear.

Reply:
You mean, I should have just one function not the two?

I have use `datetime.date.fromtimestamp` for the clinic. If I just use `date.fromtimestamp`it will complain (Parent class or module does not exist).
With the clinic declaration, this results in exactly the signature

~~~
datetime_date_fromtimestamp(PyTypeObject *type, PyObject *timestamp)
~~~

However, I cannot use this function in the `PyDateTime_CAPI` declaration (l.6075). If I try that, the compiler complains 

~~~
/home/tim/dev/cpython/Modules/_datetimemodule.c:6075:5: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
     datetime_date_fromtimestamp,
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/tim/dev/cpython/Modules/_datetimemodule.c:6075:5: note: (near initialization for ‘CAPI.Date_FromTimestamp’)
~~~

The function there is expected to have `PyObject *` as type, not `PyTypeObject *`. I don't know if that can be corrected, but reckon that it would be a breaking change to the CAPI which we do not want.

Reply:
Ah, I missed that date_fromtimestamp()` is not a new function, and it is used n other place. You could cast the function to the correct type as it was in the `date_methods` initialization:
```c
(PyCFunction)datetime_date_fromtimestamp,
```
but the current code LGTM too.